### ISSUE TITLE
Add access status to item metadata

### ISSUE THREAD

**Issue by pilasou** [DSpace/DSpace#8518]
**Current situation** 

When configured, access statuses are only available in search results and are not stored in item metadata. This prevents access status from being displayed in the item records page and being exposed through OAI PMH formats. 

**Objectives**

Add item access status to the item metadata in an automatic way. This will allow to easily display it in different context:  

- Record page in Angular UI 
- Addition to OAI formats (OpenAIRE, OAI DC) 

It can also provide the opportunity not to expose restricted item files in specific OAI formats (i.e. ORE).

This can be achieved by : 

- Using the rest endpoint for Access Status (https://github.com/DSpace/RestContract/blob/main/items.md) 
- Store the Access status value in a dc.rights metadata (the value used will be the URI of the COAR Access Rights Access Rights 1.1 (https://vocabularies.coar-repositories.org/access_rights/))

It can use a new Curation task that can then be schedule regularly or executed manually at different level (whole ; specific collections ; specific item). To avoid recalculating the access status on the whole repository, a new metadata will be created (dspace.accessstatus.time) to store the last execution dates.  

**Additional context**
This will build on https://github.com/DSpace/dspace-angular/issues/1566

**Comment by tdonohue** [comment]
@pilasou : I'm very worried about the suggestion to store item statuses in Metadata fields.  That seems like it could be complex as it'd require dynamically updating metadata field values whenever an embargo date ends or similar. Ensuring the metadata field is "synchronized" with the item status seems like it may not be easy.

Instead, I wonder if there's a way to simply display the Item status on the Item page using logic similar to how it was achieved in https://github.com/DSpace/dspace-angular/issues/1566

Similar logic could also be added to the OAI-PMH code to automatically & dynamically make the Item Status appear in the `dc.rights` field from the OAI-PMH interface ONLY. 

This is just my immediate opinion, but I think this could be much more complex than you are anticipating. We don't really have a way to auto-update metadata fields whenever an item state changes (e.g. embargo expires, file is uploaded to an item that was previously "Metadata only", permissions are changed so that the Item moves from "Open Access" -> "Restricted" or visa versa).

**Comment by pilasou** [comment]
@tdonohue : I agree that it does not offer a realtime synchronization. However, the idea is to offer the ability to create a process that can be scheduled regularly or trigger manually as admin either on a specific document or a entire collection. What we also found interesting is that when using DSPace 7 OpenAIRE submission forms, the dc.rights element is used to record COAR access status, and the OpenAIRE OAI crosswalk is using dc.rights to display access rights as required by the OpenAIRE Guidelines. The idea was also to align with this.

**Comment by tdonohue** [comment]
@pilasou : I feel the lack of real-time synchronization would make the usability poor in the UI itself...the status on the Item page would not be possible to trust, as it may not be accurate at any given time.  The status page may claim the item is "open access" when it's not. It may claim it's "embargoed" when it's not, etc.  That's not a behavior that users will be OK with.

If the goal is to simply make this status shareable in OAI-PMH, then I'd recommend looking for a way to modify the OAI-PMH Java code to dynamically read the status and index it _as if it was in_ the `dc.rights` field.  There's no strict requirement that OAI-PMH can only index metadata fields "as-is".

**Comment by paulo-graca** [comment]
@pilasou Access Status at the OAI level is already addressed with the contribution of @amgciadev and others with https://github.com/DSpace/DSpace/pull/8926

and there will be some more related developments:
https://github.com/DSpace/DSpace/issues/9075

**Comment by pilasou** [comment]
@paulo-graca Thanks! Being able to retrieve access status for bitstreams for OAI is an excellent news. Is this new dev will help pave the way to solve this other issue: https://github.com/DSpace/dspace-angular/issues/2413. I understand that the OAI part of this issue is closed, but there was another need, that do not reach consensus, on the ability to add access status to item metadata.

**Comment by paulo-graca** [comment]
For access status as a metadata in the item, I have one suggestion. I think we could extend the concept of **virtual metadata** (currently, virtual metadata only works for relationships and entities), imagine that if you could include, other configurable and calculated fields that don't belong to the item, that aren't registered in the metadatavalue table and that could extend your metadata. Currently we can do that, /full pages show us not only data that the user registered but also imported, and some other fields that don't exist in the metadavalue table like: `relation.isAuthorOfPublication` or `person.name` if the virtual metadata field is configured.
The way I see, a new virtual metadata field implementation could be created and be used to display the calculated access status for that specific item. That solution would easily enable the indexing of access status for that item, without changes.

**Comment by paulo-graca** [comment]
But it's not the ideal solution for embargoes. For embargoes, with this approach I can't see another solution unless we bring back a  script like embargo-lifter (https://wiki.lyrasis.org/display/DSDOC6x/Pre-3.0+Embargo+Lifter+Commands).

**Comment by pilasou** [comment]
@paulo-graca Why do you concider the virtual metadata not the ideal approach? Is it because it would require major dev/refactor or so? Our approach localy is not ideal either, we simply create a Curation task that is executed daily. For new items it creates the dc.rights metadata with the COAR vocabulary URL and another one (dspace.accessstatus.time) to mention the date the curation task has last updated the status. The curation task is using the Threshold parameter to set the number of days after which the status should be updated (we set it to 10 days).

**Comment by paulo-graca** [comment]
I'm not saying that the virtual metadata isn't the ideal approach. It would require some refactoring, like creating a general concept of virtual metadata and the current one being a special case of virtual metadata (for entities). I was saying that embargoes, using this approach, should have a kind of "embargo-lifter" script to update the indexed item when the start date is reached (when the status changes from embargo->open.access).

**Comment by paulo-graca** [comment]
@pilasou without any new feature I'm able to filter results by access status.
![image](https://github.com/DSpace/DSpace/assets/877488/77eae9bc-37eb-4c59-ad27-9dd9af41036d)

I'm using a custom discovery plugin. Discovery plugins let you enhance Solr document indexing at the Discovery level by performing actions that aren't available by any usual mean. My custom plugin just gets the current object access status and adds it into a Solr document field. I can create a PR with it. I also needed to change the discovery.xml config file to consider this new filter.

But the major issue still relates with embargoes. I will need to periodically run some command, like this one:
https://wiki.lyrasis.org/display/DSDOC7x/Discovery#Discovery-DiscoverySolrIndexMaintenance

to update the index of items that ended the embargo period.

**Comment by DanGastardelli** [comment]
Hi @paulo-graca, could you provide the changes you needed to make to get the filter working?

**Comment by DanGastardelli** [comment]
Hi @paulo-graca, I completely understand, thanks for the feedback.

It seems like an interesting filter, but I would need to check with @tdonohue about the PR.

**Comment by tdonohue** [comment]
All, just finding this ticket again, and it appears that this feature is also described in https://github.com/DSpace/dspace-angular/issues/2413

There is already PRs that have been created to fix that other ticket.  See https://github.com/DSpace/DSpace/pull/10294 and https://github.com/DSpace/dspace-angular/pull/3882

I have not yet reviewed or tested those PRs, but they will be considered for 9.0.  If anyone is interested in this feature, then I'd highly recommend helping test or review those PRs -- additional testers are always helpful in speeding up the acceptance process for a PR.

**Comment by paulo-graca** [comment]
@DanGastardelli @pilasou I've just added:
https://github.com/DSpace/DSpace/pull/10434

This new plugin will add the faceted filter for access status of an item.

**Comment by tdonohue** [comment]
Closing, fixed by #10434
